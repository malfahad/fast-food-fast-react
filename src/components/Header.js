import React from 'react';

const Header = () => (
  <div className="header">
    <h1 className="heading main-title"> Fast Food Fast</h1>
    <h2 className="heading"> Order your food online </h2>
  </div>
);

export default Header;
