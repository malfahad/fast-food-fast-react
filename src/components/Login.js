import React from 'react';
import { connect } from 'react-redux';
import sampleAction from '../actions/sampleAction'

class Login extends React.Component {

    render(){
        return (<div>Login form</div>)
    }
}

const mapStateToProps = state=>({
    ...state
});

export default connect(mapStateToProps,{sampleAction})(Login)