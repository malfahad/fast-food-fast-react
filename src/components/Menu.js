import React from 'react';
import { connect } from 'react-redux';
import {
  AppHeader, OrderSummary, MenuForm, AdminMenuButtons, CustomerMenuButttons,
} from './Shared';
import fetchMenuAction, { removeFromMenu, addToMenu } from '../actions/menuAction';
import { addToOrder, removeFromOrder, placeOrder } from '../actions/orderActions';

export class Menu extends React.Component {
  componentWillMount() {
    this.props.fetchMenuAction();
  }

  onRemove = e => this.props.removeFromOrder(e.target.id);

  onAdd = e => this.props.addToOrder(e.target.id);

  onRemoveFromMenu = e => this.props.removeFromMenu(e.target.id);

  onOrderSubmit = (e) => {
    e.preventDefault();
    const { orderSummary, total } = this.props.menuState;
    const items = Object.values(orderSummary).map(item => (
      `${item.count}x ${item.title} - ${item.subTotal}`
    ));
    const status = 'Created';
    const payload = { items, total, status };
    this.props.placeOrder(payload);
  }

  onMenuSubmit = (e) => {
    e.preventDefault();
    const title = e.target[0].value;
    const description = e.target[1].value;
    const amount = e.target[2].value;
    const imgUrl = e.target[3].value;
    const body = {
      title, description, amount, img_url: imgUrl,
    };
    this.props.addToMenu(body);
  }

  logout = (e) => {
    e.preventDefault();
    localStorage.removeItem('ff-token');
    localStorage.removeItem('ff-admin');
    const loginPath = (this.props.user.isAdmin) ? '/admin/login' : '/login';
    this.props.history.push(loginPath);
  }

  render() {
    const { isAdmin } = this.props.user;
    const { menu, orderSummary, total } = this.props.menuState;
    const menuIds = Object.keys(menu);

    const menuItems = menuIds.map(key => (
      <div className="menu-item" key={key}>
        <img className="menu-item-img" src={menu[key].image_url} alt="menu item" />
        <h4 className="heading menu-item-title">
          {' '}
          {menu[key].title}
          {' '}
        </h4>
        <p>
          {' '}
          {menu[key].description}
          {' '}
        </p>
        <p className="item-price">
UGX
          {' '}

          {menu[key].amount}

        </p>
        {isAdmin ? (
          <AdminMenuButtons
            id={key}
            onRemoveFromMenu={this.onRemoveFromMenu}
          />
        ) : (
          <CustomerMenuButttons
            id={key}
            count={Object.prototype.hasOwnProperty.call(orderSummary, key)
              ? orderSummary[key].count : 0}
            onRemove={this.onRemove}
            onAdd={this.onAdd}
          />
        )}

      </div>
    ));

    const leftPane = isAdmin
      ? (
        <MenuForm
          onSubmit={this.onMenuSubmit}
        />
      ) : (
        <OrderSummary
          onOrderSubmit={this.onOrderSubmit}
          orderSummary={Object.values(orderSummary)}
          total={total}
        />
      );

    return (
      <div>
        <AppHeader logout={this.logout} />
        <div className="container">
          {leftPane}
          <div className="column-main">
            <h3 className="heading">Restaurant Menu</h3>
            <p> This is the current fast food fast items menu.</p>
            <div id="menu">
              {menuItems}
            </div>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  user: state.userReducer,
  menuState: state.menuReducer,
});

export default connect(mapStateToProps, {
  fetchMenuAction,
  addToOrder,
  removeFromOrder,
  removeFromMenu,
  addToMenu,
  placeOrder,
})(Menu);
