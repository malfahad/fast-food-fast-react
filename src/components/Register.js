import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import authAction from '../actions/authAction';

class Register extends React.Component {
  componentWillReceiveProps(nextProps) {
    if (nextProps.user.isLoggedIn) {
      localStorage.setItem('ff-token', nextProps.user.authorization);
      this.props.history.push('/');
    }
  }

    handleSubmit = (e) => {
      e.preventDefault();
      const fullName = e.target[0].value;
      const email = e.target[1].value;
      const password = e.target[2].value;
      const payload = {
        'full name': fullName,
        email,
        password,
      };
      this.props.authAction(false, payload);
    }

    render() {
      return (
        <div>
          <div className="header">
            <h1 className="heading main-title"> Fast Food Fast</h1>
            <h2 className="heading"> Order your food online </h2>
          </div>

          <div className="content">

            <form id="form-user-signup" onSubmit={this.handleSubmit} className="form-login" action="/">
              <h5 className="heading">Register a new account </h5>
              <p id="server-error" className="server-error">{this.props.user.authErrors}</p>
              <div className="form-group">
                <input type="text" className="form-control" placeholder="Full name" id="full_name" required />
              </div>
              <div className="form-group">
                <input type="email" className="form-control" placeholder="Email" id="email" required />
              </div>
              <div className="form-group">
                <input type="password" className="form-control" placeholder="Password" id="password" required />
              </div>
              <div className="form-group">
                <input type="password" className="form-control" placeholder="Repeat password" id="password2" required />
              </div>
              <div className="form-group">
                <input type="submit" className="form-control submit" value="Sign up" id="submit" />
              </div>
              <p>
                {' '}
Already registered?
                <Link to="/login">Sign in</Link>
              </p>

            </form>
          </div>
        </div>

      );
    }
}

const mapStateToProps = state => ({
  user: state.userReducer,
});

export default connect(mapStateToProps, { authAction })(Register);
