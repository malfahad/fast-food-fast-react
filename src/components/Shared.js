import React from 'react';
import { Link } from 'react-router-dom';

const Header = () => (
  <div className="header">
    <h1 className="heading main-title"> Fast Food Fast</h1>
    <h2 className="heading"> Order your food online </h2>
  </div>
);

export const AppHeader = props => (
  <div>
    <div className="heading-main">
      <h1 className="heading">Fast Food Fast</h1>
      <h3 className="heading">{props.title}</h3>
      <ul>
        <li className="btn btn-default logout"><a id="user-logout" href="#" onClick={props.logout}>Logout</a></li>
        <li className="btn btn-default logout"><Link to="/orders">Order history</Link></li>
        <li className="btn btn-default logout"><Link to="/">Menu</Link></li>
      </ul>
    </div>

    <div className="gap" />

  </div>
);

export const OrderItem = props => (
  <div className="menu-item">
    <img className="menu-item-img" alt="order item" src="https://cdn2.iconfinder.com/data/icons/bar-2/64/padnote-order-restaurant-writing_tool-512.png" />
    <h4 className="heading menu-item-title"> Order number #100341 </h4>
    <ul>
      {props.itemList.map(item => <li>{item}</li>)}
    </ul>
    <p>
Total:
      {' '}
      {' '}
      <span>
Ush
        {props.total}
      </span>


    </p>
    <p>
Order Status:
      {' '}
      <span className="order-status completed">{props.status}</span>


    </p>
  </div>
);

export const OrderSummary = props => (
  <div className="column-small persist">
    <h3 className="heading">Order Summary </h3>
    <ul id="order-summary">
      {props.orderSummary.map(item => (
        <li key={`${item.title}${item.count}`}>{`${item.count}x ${item.title}  - ${item.subTotal}`}</li>
      ))}
    </ul>
    <h4 id="order-total">{props.total > 0 ? `Total Ush ${props.total}` : ''}</h4>
    <button id="order-submit" type="button" onClick={props.onOrderSubmit} className="menu-item-button" hidden={props.total === 0}>Submit Order</button>
  </div>
);

export const MenuForm = props => (
  <div className="column-small">
    <h3 className="heading">Add menu Item </h3>
    <form id="form-menu-item" onSubmit={props.onSubmit} action="/" encType="application/x-www-form-urlencoded">
      <p className="server-error" id="server-error" />
      <div className="form-group2 ">
        <input type="text" className="form-control2" placeholder="Menu Item Title" required id="menu-item-title" />
      </div>
      <div className="form-group2 ">
        <textarea type="text" className="form-control2" placeholder="Menu Item description" required id="menu-item-desc" />
      </div>
      <div className="form-group2 ">
        <input type="number" className="form-control2" placeholder="Amount" required id="menu-item-amount" />
      </div>
      <div className="form-group2 ">
        <small><label htmlFor="menu-item-img"> Menu Item Image: </label></small>
        <br />
        <input type="text" className="form-control2" placeholder="Menu Item Image url" id="menu-item-img" />
      </div>
      <div className="form-group2">
        <input type="submit" className="form-control submit" value="Create new" id="submit" />
      </div>
    </form>
  </div>
);

export const CustomerMenuButttons = props => (
  <div>
    <p className="item-counter">{props.count > 0 ? `${props.count} added` : ''}</p>
    <button id={props.id} type="button" className="menu-item-button" onClick={props.onAdd} hidden={props.count >= 10}>Add</button>
    <button id={props.id} type="button" className="menu-item-button" onClick={props.onRemove} hidden={props.count < 1}>Remove</button>
  </div>
);
export const AdminMenuButtons = props => (
  <div>
    <button id={props.id} type="button" className="menu-item-button" onClick={props.onRemoveFromMenu}>Remove from Menu</button>
  </div>
);


export default Header;
