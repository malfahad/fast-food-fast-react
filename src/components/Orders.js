import React from 'react';
import { connect } from 'react-redux';
import { fetchOrders } from '../actions/orderActions';
import { AppHeader, OrderItem } from './Shared';

export class Orders extends React.Component {
  componentWillMount() {
    this.props.fetchOrders();
  }

    logout = (e) => {
      e.preventDefault();
      localStorage.removeItem('ff-token');
      const loginPath = (this.props.user.isAdmin) ? '/admin/login' : '/login';
      this.props.history.push(loginPath);
    }

    render() {
      const { orderHistory } = this.props;
      const orders = Object.values(orderHistory);
      orders.reverse();
      const orderItems = orders.length < 1 ? <p>No past orders</p> : (
        <div>
          {
          orders.map(order => (
            <OrderItem
              status={order.status}
              total={order.total}
              itemList={order.items}
            />
          ))
        }
        </div>
      );

      return (
        <div>
          <AppHeader
            logout={this.logout}
            title="order History"
          />
          <div className="container">
            <div className="column-wide">
              <h3 className="heading">My Order history</h3>
              <div id="orders-list">
                {orderItems}
              </div>
            </div>
          </div>
        </div>
      );
    }
}


const mapStateToProps = state => ({
  user: state.userReducer,
  menuState: state.menuReducer,
  orderHistory: state.ordersReducer.orders,
});

export default connect(mapStateToProps, {
  fetchOrders,
})(Orders);
