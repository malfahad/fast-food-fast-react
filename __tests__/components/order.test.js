import React from 'react';
import { shallow } from 'enzyme';
import { Orders } from '../../src/components/Orders';

describe('<Orders/>', () => {
  it('should render empty order list without crashing', () => {
    const props = {
      fetchOrders: jest.fn(),
      orderHistory: {},
      user: {
        isAdmin: false,
      },
    };
    const wrapper = shallow(<Orders {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('should render orders list without crashing', () => {
    const props = {
      fetchOrders: jest.fn(),
      orderHistory: {
        1: {
          items: ['2x Fries - Ush 2000'],
          status: 'Created',
          total: 2000,
        },
      },
      user: {
        isAdmin: false,
      },
    };
    const wrapper = shallow(<Orders {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should call updateorderStatus on chage selection', () => {
    const props = {
      fetchOrders: jest.fn(),
      updateOrderStatus: jest.fn(),
      orderHistory: {
        8: {
          items: ['2x Fries - Ush 2000'],
          status: 'Created',
          total: 2000,
        },
      },
      user: {
        isAdmin: false,
      },
    };
    const event = {
      preventDefault: jest.fn(),
      target: {
        1: { value: 'Confirmed', id: 8 },
        2: { value: 'Rejected', id: 8 },
        selectedIndex: 1,
      },

    };
    const wrapper = shallow(<Orders {...props} />);
    wrapper.instance().onStatus(event);
    expect(wrapper.instance().props.updateOrderStatus).toHaveBeenCalled();
  });


  it('should navigato to admin login on admin logout', () => {
    const props = {
      fetchOrders: jest.fn(),
      user: {
        isAdmin: true,
      },
      orderHistory: {},
      history: {
        push: jest.fn(),
      },
    };
    const event = {
      preventDefault: jest.fn(),
    };
    const wrapper = shallow(<Orders {...props} />);
    wrapper.instance().logout(event);
    expect(wrapper.instance().props.history.push).toHaveBeenCalled();
  });


  it('should navigato to user login on user logout', () => {
    const props = {
      fetchOrders: jest.fn(),
      user: {
        isAdmin: false,
      },
      orderHistory: {},
      history: {
        push: jest.fn(),
      },
    };
    const event = {
      preventDefault: jest.fn(),
    };
    const wrapper = shallow(<Orders {...props} />);
    wrapper.instance().logout(event);
    expect(wrapper.instance().props.history.push).toHaveBeenCalled();
  });
});
