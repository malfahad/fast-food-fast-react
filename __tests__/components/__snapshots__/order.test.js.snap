// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Orders/> should render empty order list without crashing 1`] = `ShallowWrapper {}`;

exports[`<Orders/> should render orders list without crashing 1`] = `ShallowWrapper {}`;
